File: src/ASTComponents/ExpressionComponents/prefix_expression.rb

class PreFixExpression
    def initialize(token, operator, right_exp)
        @operator = operator
        @rhs_exp = right_exp
        @token = token
    end

    def visit(semantic_analyzer)
        semantic_analyzer.analyze_node_locally(self)
    end

    def getRhsExp()
        @rhs_exp
    end

    def getOperator()
        @operator
    end

    def getName()
        @token
    end

    def _printLiteral(repr_list)
        repr_list.append("(")
        @rhs_exp._printLiteral(repr_list)
        repr_list.append(")")
    end

    def _printTokType(type_list)
        type_list.append("(")
        @rhs_exp._printLiteral(type_list)
        type_list.append(")")
    end

    def toJSON()
        return {
            "type" => "prefix",
            "token" => {
                "literal" => @token.getText(),
                "type" => @token.getType(),
                "line_number" => @token.getLine()
            },
            "rhs_exp" => @rhs_exp.toJSON()
        }
    end
end

File: src/ASTComponents/ExpressionComponents/prefix_expression.py



class PrefixExpression:
    def __init__(self):
        self.token = None
        self.rhs_exp = None

    def add_name(self, token):
        self.token = token

    def get_name(self):
        return self.token

    def add_rhs_exp(self, rhs_expression):
        self.rhs_exp = rhs_expression

    def get_rhs_exp(self):
        return self.rhs_exp

    def accept(self, expression_analyzer):
        expression_analyzer.visit_prefix_expression(self)

    def has_left_expression(self):
        return False
    
    def has_right_expression(self):
        return self.rhs_exp is not None
